src/xcode/ENA/ENA/Source/Scenes/QRScanner/QRScannerCoordinator.swift

```diff
@@ -135,34 +144,78 @@ class QRScannerCoordinator {
 	}
 
 	private func showQRCodeResult(qrCodeResult: QRCodeResult) {
-		parentViewController?.dismiss(animated: true, completion: { [weak self] in
-			switch qrCodeResult {
-			case let .coronaTest(testRegistrationInformation):
-				self?.showScannedTestResult(testRegistrationInformation)
-			case let .certificate(certificateResult):
-				self?.showScannedHealthCertificate(certificateResult)
-			case let .traceLocation(traceLocation):
-				self?.showScannedCheckin(traceLocation)
-			}
-		})
+		switch qrCodeResult {
+		case let .coronaTest(testRegistrationInformation):
+			showScannedTestResult(testRegistrationInformation)
+		case let .certificate(certificateResult):
+			showScannedHealthCertificate(certificateResult)
+		case let .traceLocation(traceLocation):
+			showScannedCheckin(traceLocation)
+		}
 	}
 
+	// swiftlint:disable cyclomatic_complexity
 	private func showScannedTestResult(
 		_ testRegistrationInformation: CoronaTestRegistrationInformation
 	) {
-		switch presenter {
-		case .submissionFlow:
-			didScanCoronaTestInSubmissionFlow?(testRegistrationInformation)
-		case .onBehalfFlow:
-			let parentPresentingViewController = parentViewController?.presentingViewController
+		let recycleBinItemToRestore: RecycleBinItem?
+		switch testRegistrationInformation {
+		case .pcr(guid: _, qrCodeHash: let qrCodeHash),
+			.antigen(qrCodeInformation: _, qrCodeHash: let qrCodeHash):
+			recycleBinItemToRestore = store.recycleBinItems.first {
+				guard case .coronaTest(let coronaTest) = $0.item else {
+					return false
+				}
 
-			parentViewController?.dismiss(animated: true) {
-				self.parentViewController = parentPresentingViewController
+				return coronaTest.qrCodeHash == qrCodeHash
+			}
+		case .teleTAN:
+			recycleBinItemToRestore = nil
+		}
+
+		if let recycleBinItemToRestore = recycleBinItemToRestore {
+			showTestRestoredFromBinAlert(recycleBinItem: recycleBinItemToRestore)
+			return
+		}
+
+		qrScannerViewController?.dismiss(animated: true) { [weak self] in
+			guard let self = self else { return }
+
+			switch self.presenter {
```

Comment:
I wonder if the logic in this switch shouldn't be the same which is called after `showTestRestoredFromBinAlert` is finished. Do we have a redundant implementation?

Reply:
The details are different, I think this is not easy to generalize. E.g. in this case we start the coordinator with `CoronaTestRegistrationInformation`, in `showTestRestoredFromBinAlert` the coordinator is started with the `CoronaTestType`.